ipni-datalog.md

```diff
@@ -0,0 +1,136 @@
+
+2. Anyone could publish an IPNI advertisement that will associate `block..1` to some `who..knows` multihash which may not be a [DAG Index], yet client will only discover that after fetching and trying to parse as such.
+
+3. Client can not query specific relations, meaning fact that `block..1` is block of the DAG and fact that `blb...left` is a blob containing DAG blocks is not captured in any way.
+
+4. We also have not captured relation between `blob..left` and `block..1` in any way other than they both relate to `bag...idx`.
+
+## Proposal
+
+Leverage [IPNI] such that we could resolve `entity` that [multihash] relates to
+without having to resolve and fetch [DAG Index] first.
+
+General idea is to capture relation names in a way that it could be utilized during lookups. This also would make indexing protocol open-ended and extensible by anyone without having intermediaries like W3Up having to adopt those extensions.
+
+To illustrate this lets derive another lookup table from the same [DAG Index][index example], however in this case we will replace `key` column with pair of `attribute` (describing relation to an `entity`) and `value` which is a [multihash] related to `entity`.
+
+| entity      | attribute             | value       |
+| ----------- | --------------------- | ------------|
+| `blb..left` | `dag@0.1/shard`       | `bafy..dag` |
+| `blb..left` | `dag@0.1/shard/slice` | `block..1`  |
+| `blb..left` | `dag@0.1/shard/slice` | `block..2`  |
+| `blb..left` | `dag@0.1/shard/slice` | `block..3`  |
+| `blb..left` | `dag@0.1/shard/slice` | `block..4`  |
+| `blb..right`| `dag@0.1/shard`       | `bafy..dag` |
+| `blb..right`| `dag@0.1/shard/slice` | `block..5`  |
+| `blb..right`| `dag@0.1/shard/slice` | `block..6`  |
+| `blb..right`| `dag@0.1/shard/slice` | `block..7`  |
+| `blb..right`| `dag@0.1/shard/slice` | `block..8`  |
+
+In this variant we have virtual `key` that is derived from `attribute` and `value` components in a deterministic manner e.g. by concatenating their bytes and computing a [multihash].
+
+This means that we could find shards of he `bafy..dag` via IPNI query that resolves `toKey(['bafy..dag', 'dag@0.1/shard'])` key.
+
+If system drops `blob..left` will drop all the records associated with that entity as it would be a single IPNI advertisement.
+
+Because we get shards from single query we do not need to fetch [DAG Index] to start resolving location commitments. We could start resolving them right away and we gate IPNI with something like GraphQL we could even resolve those in a single roundtrip and cache for subsequent queries.
+
+Location commitments can be also indexed and advertised on [IPNI] in a very similar way.
+
+| entity   | attribute             | value        |
+| ---------|-----------------------|--------------|
+| `loc..1` | `assert/location@0.2` | `bafy..left` |
+| `loc..2` | `assert/location@0.2` | `bafy..left` |
+| `loc..3` | `assert/location@0.2` | `bafy..right`|
```

Comment:
So, resolving might look like:

1. Concurrently
    1. Query IPNI for `bafy..dag`
        * Receive `bag...idx`
        * Fetch `bag...idx`
    1. Query IPNI for `toKey(['bafy..dag', 'dag@0.1/shard'])`
        * Receive `blb..left`, `blb..right`
        * Concurrently
            1. Query IPNI for `toKey(['blb..left', 'assert/location@0.2'])`
                * Receive `loc..1`, `loc..2`
            1. Query IPNI for `toKey(['blb..right', 'assert/location@0.2'])`
                * Receive `loc..3`
1. Export DAG with all required information

---

@Gozala I'm just wondering if `dag@0.1/shard` should be something more like `assert/parition@0.2` and `dag@0.1/shard/slice` be `assert/inclusion@0.2`.

Reply:
Yeah I think names could be improved I just copied paths from the DAG Index keys

Reply:
> So, resolving might look like:
> 
> 1. Concurrently
>    
>    1. Query IPNI for `bafy..dag`
>       
>       * Receive `bag...idx`
>       * Fetch `bag...idx`
>    2. Query IPNI for `toKey(['bafy..dag', 'dag@0.1/shard'])`
>       
>       * Receive `blb..left`, `blb..right`
>       * Concurrently
>         
>         1. Query IPNI for `toKey(['blb..left', 'assert/location@0.2'])`
>            
>            * Receive `loc..1`, `loc..2`
>         2. Query IPNI for `toKey(['blb..right', 'assert/location@0.2'])`
>            
>            * Receive `loc..3`
> 2. Export DAG with all required information

Turns out I have not pushed some changes (my apologies) which have mislead you. Here are the actual steps that would be required for computing `bafy..dag` DAG

- Query IPNI for the `bafy..dag` blobs `toKey(['dag@0.1/shard', 'bafy..dag'])`
    - Receive `[blb..left, `blb..right]`
    - Concurrently query IPNI for
       - Blob locations `toKey(['assert/location@0.2', 'blb..left'])`
          - Fetch blob from location
       - Blob partition `toKey(['blob/partition@0.2', 'blb..left'])`
          - Fetch resolved partition index
       - Blob locations `toKey(['assert/location@0.2', 'blb..right'])`
          - Fetch blob from location
       - Blob partition `toKey(['blob/partition@0.2', 'blb..right'])`
          - Fetch resolved partition index
     - Slice up fetched blob to blocks per partition index & put together a DAG

Reply:
And I want to stress again roundtrips can be reduced to 1 if we put something in front of IPNI that can perform a join